### PR TITLE
[Fix #8684] Fix an error for `Lint/InterpolationCheck` cop

### LINKED ISSUE #8684: Error in Lint/InterpolationCheck
## Expected behavior

Cop runs w/o error

## Actual behavior

Error is thrown when scanning:

```
An error occurred while Lint/InterpolationCheck cop was inspecting /Users/tsmith/dev/the_world/L7-mysql/providers/server.rb:25:42.
Expected a Parser::Source::Range, Comment or Rubocop::AST::Node, got NilClass
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/corrector.rb:92:in `to_range'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/corrector.rb:101:in `check_range_validity'
/usr/local/lib/ruby/gems/2.7.0/gems/parser-2.7.1.4/lib/parser/source/tree_rewriter.rb:371:in `combine'
/usr/local/lib/ruby/gems/2.7.0/gems/parser-2.7.1.4/lib/parser/source/tree_rewriter.rb:194:in `replace'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/lint/interpolation_check.rb:45:in `autocorrect'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/lint/interpolation_check.rb:32:in `block in on_str'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/base.rb:333:in `correct'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/base.rb:126:in `add_offense'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/lint/interpolation_check.rb:31:in `on_str'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/commissioner.rb:99:in `block (2 levels) in trigger_responding_cops'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/commissioner.rb:152:in `with_cop_error_handling'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/commissioner.rb:98:in `block in trigger_responding_cops'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/commissioner.rb:97:in `each'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/commissioner.rb:97:in `trigger_responding_cops'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/commissioner.rb:70:in `on_str'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-ast-0.3.0/lib/rubocop/ast/traversal.rb:59:in `block in on_array'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-ast-0.3.0/lib/rubocop/ast/traversal.rb:59:in `each'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-ast-0.3.0/lib/rubocop/ast/traversal.rb:59:in `on_array'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/commissioner.rb:72:in `on_array'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-ast-0.3.0/lib/rubocop/ast/traversal.rb:116:in `block in on_send'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-ast-0.3.0/lib/rubocop/ast/traversal.rb:113:in `each'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-ast-0.3.0/lib/rubocop/ast/traversal.rb:113:in `each_with_index'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-ast-0.3.0/lib/rubocop/ast/traversal.rb:113:in `on_send'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/commissioner.rb:72:in `on_send'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-ast-0.3.0/lib/rubocop/ast/traversal.rb:169:in `on_block'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/commissioner.rb:72:in `on_block'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-ast-0.3.0/lib/rubocop/ast/traversal.rb:59:in `block in on_begin'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-ast-0.3.0/lib/rubocop/ast/traversal.rb:59:in `each'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-ast-0.3.0/lib/rubocop/ast/traversal.rb:59:in `on_begin'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/commissioner.rb:72:in `on_begin'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-ast-0.3.0/lib/rubocop/ast/traversal.rb:173:in `on_block'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/commissioner.rb:72:in `on_block'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-ast-0.3.0/lib/rubocop/ast/traversal.rb:59:in `block in on_begin'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-ast-0.3.0/lib/rubocop/ast/traversal.rb:59:in `each'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-ast-0.3.0/lib/rubocop/ast/traversal.rb:59:in `on_begin'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/commissioner.rb:72:in `on_begin'
/usr/local/lib/ruby/gems/2.7.0/gems/rubocop-ast-0.3.0/lib/rubocop/ast/traversal.rb:14:in `walk'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/commissioner.rb:85:in `investigate'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/team.rb:152:in `investigate_partial'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cop/team.rb:83:in `investigate'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/runner.rb:299:in `inspect_file'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/runner.rb:249:in `block in do_inspection_loop'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/runner.rb:281:in `block in iterate_until_no_changes'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/runner.rb:274:in `loop'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/runner.rb:274:in `iterate_until_no_changes'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/runner.rb:245:in `do_inspection_loop'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/runner.rb:125:in `block in file_offenses'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/runner.rb:150:in `file_offense_cache'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/runner.rb:124:in `file_offenses'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/runner.rb:115:in `process_file'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/runner.rb:94:in `block in each_inspected_file'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/runner.rb:93:in `each'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/runner.rb:93:in `reduce'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/runner.rb:93:in `each_inspected_file'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/runner.rb:79:in `inspect_files'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/runner.rb:40:in `run'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cli/command/execute_runner.rb:22:in `execute_runner'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cli/command/execute_runner.rb:14:in `run'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cli/command.rb:11:in `run'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cli/environment.rb:18:in `run'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cli.rb:65:in `run_command'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cli.rb:72:in `execute_runners'
/Users/tsmith/dev/oss/rubocop/lib/rubocop/cli.rb:41:in `run'
/Users/tsmith/dev/oss/rubocop/exe/rubocop:13:in `block in <top (required)>'
/usr/local/Cellar/ruby/2.7.1_2/lib/ruby/2.7.0/benchmark.rb:308:in `realtime'
/Users/tsmith/dev/oss/rubocop/exe/rubocop:12:in `<top (required)>'
/usr/local/lib/ruby/gems/2.7.0/bin/rubocop:23:in `load'
/usr/local/lib/ruby/gems/2.7.0/bin/rubocop:23:in `<top (required)>'
/usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:63:in `load'
/usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:63:in `kernel_load'
/usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:28:in `run'
/usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/cli.rb:476:in `exec'
/usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor.rb:399:in `dispatch'
/usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/cli.rb:30:in `dispatch'
/usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/base.rb:476:in `start'
/usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/cli.rb:24:in `start'
/usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/exe/bundle:46:in `block in <top (required)>'
/usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/lib/bundler/friendly_errors.rb:123:in `with_friendly_errors'
/usr/local/lib/ruby/gems/2.7.0/gems/bundler-2.1.4/exe/bundle:34:in `<top (required)>'
/usr/local/opt/ruby/bin/bundle:23:in `load'
/usr/local/opt/ruby/bin/bundle:23:in `<main>'
```

## Steps to reproduce the problem

Scan this ruby:

```ruby
  %w("#{new_resource.name}-binlogcleaner" "#{new_resource.name}-monitoring").each do |cron|
    cron_d cron do
      action :delete
    end
  end
```

## RuboCop version
Master

```
0.90.0 (using Parser 2.7.1.4, rubocop-ast 0.3.0, running on ruby 2.7.1 x86_64-darwin19)
```

### PR DESCRIPTION
**Pull request by tejasbubane**
Closes https://github.com/rubocop-hq/rubocop/issues/8684

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](https://github.com/rubocop-hq/rubocop/blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop-hq/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Run `bundle exec rake default`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/lint/interpolation_check.rb [hunk 1]
```diff
@@ -35,6 +35,10 @@ def on_str(node)
 
         private
 
+        def regex_type?(parent)
+          parent && (parent.dstr_type? || parent.regexp_type?)
+        end
```
Comment (tejasbubane) [inline lib/rubocop/cop/lint/interpolation_check.rb:40]: This is to overcome the method complexity offense raised by rubocop

Comment (rrosenblum) [reply lib/rubocop/cop/lint/interpolation_check.rb:40]: Hmm, I think we can use safe navigation to simplify this a little bit

```ruby
def regex_type?(parent)
  parent&.dstr_type? || parent&.regexp_type?
end
```

**Review by rrosenblum — APPROVED** [review]
Change looks good to me

File: lib/rubocop/cop/lint/interpolation_check.rb [hunk 2]
```diff
@@ -35,6 +35,10 @@ def on_str(node)
 
         private
 
+        def regex_type?(parent)
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/interpolation_check.rb:38]: In this context `node` is a better name.

File: lib/rubocop/cop/lint/interpolation_check.rb [hunk 3]
```diff
@@ -35,6 +35,10 @@ def on_str(node)
...
+          parent&.dstr_type? || parent&.regexp_type?
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/interpolation_check.rb:39]: I don't quite get how those two checks imply that something's a regexp type. I'd say the method name is quite confusing.

**Comment by tas50** [comment]
Any chance we'll see this one merged for the next release?

**Comment by marcandre** [comment]
ping @tejasbubane

**Comment by tejasbubane** [comment]
Review comments fixed and rebased on master.

File: lib/rubocop/cop/lint/interpolation_check.rb [hunk 4]
```diff
@@ -23,10 +23,10 @@ class InterpolationCheck < Base
               'Use double quoted strings if you need interpolation.'
 
         def on_str(node)
-          parent = node.parent
-          return if parent && (parent.dstr_type? || parent.regexp_type?)
+          return if string_or_regex?(node.parent)
           return unless /(?<!\\)#\{.*\}/.match?(node.source)
           return if heredoc?(node)
+          return unless node.loc.begin && node.loc.end
```
Comment (bbatsov) [inline lib/rubocop/cop/lint/interpolation_check.rb:29]: You might want to add a comment here explaining the need for this or extract into a method with a descriptive name.

**Comment by bbatsov** [comment]
@tejasbubane Apart from my other remark - this needs to be rebased on top of `master` again.

**Comment by marcandre** [comment]
> rebased on top of `master` again.

And check out `rake changelog:*` for next PRs :-)

**Comment by bbatsov** [comment]
@tejasbubane ping :-)

File: lib/rubocop/cop/lint/interpolation_check.rb (hunk 4 above)
Comment (tejasbubane) [reply lib/rubocop/cop/lint/interpolation_check.rb:30]: Changed.

**Comment by tejasbubane** [comment]
Changes done here. Sorry for the delay.

**Comment by bbatsov** [comment]
Thanks!

**Pull request merged**